### PR TITLE
WIP: Reject double-spends of shielded nullifiers and UTXOs

### LINKED ISSUE #2230: Stop double-spends by checking nullifiers in the finalized state
## Motivation

Zebra needs to check the double-spend consensus rule for shielded spends in the finalized state.

#### Nullifiers

> A nullifier MUST NOT repeat either within a transaction, or across transactions in a valid blockchain. Sprout and Sapling and Orchard nullifiers are considered disjoint, even if they have the same bit pattern.

https://zips.z.cash/protocol/nu5.pdf#nullifierset

> A transaction is not valid if it would have added a nullifier to the nullifier set that already exists in the set

https://zips.z.cash/protocol/nu5.pdf#commitmentsandnullifiers

### Existing Design

> **`pub(super) fn queue_and_commit_finalized_blocks(&mut self, queued_block: QueuedBlock)`**
> ...
> Iterate over the enumerated transactions in the block. For each transaction:
> ...
> For each `JoinSplit` description in the transaction, insert `(nullifiers[0],())` and `(nullifiers[1],())` into `sprout_nullifiers`.
> For each `Spend` description in the transaction, insert `(nullifier,())` into `sapling_nullifiers`.
> For each `Orchard` description in the transaction, insert `(nullifier,())` into `orchard_nullifiers`.

https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0005-state-updates.md#pubsuper-fn-queue_and_commit_finalized_blocksmut-self-queued_block-queuedblock

### Design Change

- [x] Inserts into RocksDB column families should return an error if the key already exists.

This design change enforces the double-spend rules for each shielded pool.

It also enforces single writes of the other 8 column families in the data structure (as of 1 June 2021):
https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0005-state-updates.md#rocksdb-data-structures

### Implementation Change

- [x] Make sure the existing batch write code propagates errors to the validator
- [x] Make sure duplicate spends of UTXOs are rejected by the finalized state
- [x] Implement a RocksDB merge operator that rejects overwriting existing values
  - https://docs.rs/rocksdb/0.16.0/rocksdb/merge_operator/index.html
  - This is a C callback, so we can't panic in this function.
- [x] Apply it to every column family in `FinalizedState::new` using `set_merge_operator_associative`
  - https://docs.rs/rocksdb/0.16.0/rocksdb/struct.Options.html#method.set_merge_operator_associative
- [ ] Generate useful error messages

Here's the RocksDB `merge` feature documentation:
https://github.com/facebook/rocksdb/wiki/Merge-Operator#what

### Testing

- [ ] Test that duplicate sprout, sapling, and orchard nullifiers are rejected by the finalized state
- [ ] Test that duplicate spends of UTXOs are rejected by the finalized state

Optional:
- [ ] Proptest the merge operator function

## Alternatives

We could try to fetch each value before writing it, but that's error-prone. We'd probably forget to do the check somewhere in the code.

### LINKED ISSUE #2231: Stop double-spends by checking nullifiers and UTXO spends in each non-finalized chain
## Motivation

Zebra needs to check the double-spend consensus rule for shielded and transparent inputs, in each non-finalized chain.

#### Nullifiers

> A nullifier MUST NOT repeat either within a transaction, or across transactions in a valid blockchain. Sprout and Sapling and Orchard nullifiers are considered disjoint, even if they have the same bit pattern.

https://zips.z.cash/protocol/nu5.pdf#nullifierset

> A transaction is not valid if it would have added a nullifier to the nullifier set that already exists in the set

https://zips.z.cash/protocol/nu5.pdf#commitmentsandnullifiers

> A note can change from being unspent to spent as a node’s view of the best valid block chain is extended by new transactions. Also, block chain reorganizations can cause a node to switch to a different best valid block chain that does not contain the transaction in which a note was output

https://zips.z.cash/protocol/nu5.pdf#decryptivk

Note: reorganizations can also change the best chain to one where a note was spent

#### UTXO spends

> Validation state associated with transparent inputs and outputs, such as the UTXO (Unspent Transaction Output) set, is not described in this document; it is used in essentially the same way as in Bitcoin.

https://zips.z.cash/protocol/protocol.pdf#transactions

> each output of a particular transaction can only be used as an input once in the block chain.
> Any subsequent reference is a forbidden double spend - an attempt to spend the same satoshis twice.

https://developer.bitcoin.org/devguide/block_chain.html#introduction

> Any input within this block can spend an output which also appears in this block (assuming the spend is otherwise valid).
> However, the TXID corresponding to the output must be placed at some point before the TXID corresponding to the input.
> This ensures that any program parsing block chain transactions linearly will encounter each output before it is used as an input.

https://developer.bitcoin.org/reference/block_chain.html#merkle-trees

### Existing Design

> **`pub fn push(&mut self, block: Arc<Block>)`**
> ...
> Add nullifiers to the appropriate `self.<version>_nullifiers`

https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0005-state-updates.md#pub-fn-pushmut-self-block-arcblock

### Design Change

- [ ] Inserts into `Chain.<version>_nullifiers` should return an error if the key already exists
- [ ] Before adding a block to a non-finalized chain, check that its nullifiers are not in the finalized state

This design change enforces the double-spend rules for each shielded pool.

- [ ] Check transparent spends against previous transactions in the block, the relevant non-finalized chain, and the finalized state

### Testing

- [ ] Test that duplicate sprout, sapling, and orchard nullifiers are rejected by the non-finalized state
- [ ] Test that duplicate transparent UTXOs are rejected by the non-finalized state

## Alternatives

This is a required consensus rule. There aren't really any alternatives to returning an error here.

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

Zebra needs to check the double-spend consensus rule for shielded spends in the finalized state.

### Specifications

See #2231

### Designs

The state RFC is updated by this PR:
https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0005-state-updates.md#rocksdb-data-structures

## Solution

Closes #2231.

Before adding a block to a non-finalized chain:
- [ ] Check for nullifier double-spends 
- [ ] Check for UTXO double-spends

### Testing

- [x] Test that duplicate sprout nullifiers are rejected
- [ ] Test that duplicate sapling and orchard nullifiers are rejected
- [x] Test that duplicate spends of UTXOs are rejected

I've started a cached state test for commit be39530 here:
https://github.com/ZcashFoundation/zebra/actions/runs/985595375

## Review

@jvff can do the initial review on this PR.

It's somewhat urgent, because the design and code changes will conflict with the new column families for anchors, history, and value pools.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

File: zebra-state/src/service/finalized_state.rs [hunk 1]
```diff
@@ -39,16 +43,36 @@ pub struct FinalizedState {
 impl FinalizedState {
     pub fn new(config: &Config, network: Network) -> Self {
         let (path, db_options) = config.db_config(network);
+
+        // Reject updates to existing values in column families
+        // with the following Merge operators:
```
Comment (teor2345) [inline zebra-state/src/service/finalized_state.rs]:
```suggestion
        // with the following Merge rules:
```

File: book/src/dev/rfcs/0005-state-updates.md [hunk 2]
```diff
@@ -600,20 +600,39 @@ order on byte strings is the numeric ordering).
...
+| `utxo_by_outpoint`    | `OutPoint`            | `transparent::Output`                | Delete |
+| `sprout_nullifiers`   | `sprout::Nullifier`   | `()`                                 | Never  |
+| `sapling_nullifiers`  | `sapling::Nullifier`  | `()`                                 | Never  |
+| `orchard_nullifiers`  | `orchard::Nullifier`  | `()`                                 | Never  |
+| `sprout_anchors`      | `sprout::tree::Root`  | `()`                                 | Never  |
+| `sprout_incremental`  | *TODO*                | *TODO*                               | *TODO* |
+| `sapling_anchors`     | `sapling::tree::Root` | `()`                                 | Never  |
+| `sapling_incremental` | *TODO*                | *TODO*                               | *TODO* |
+| `orchard_anchors`     | `orchard::tree::Root` | `()`                                 | Never  |
+| `orchard_incremental` | *TODO*                | *TODO*                               | *TODO* |
```
Comment (teor2345) [inline book/src/dev/rfcs/0005-state-updates.md]: @dconnolly will we ever want to insert the same anchor twice?
What happens if there are no transactions for a particular pool in a block?

Either way is fine, I just need to update the design here.

File: zebra-state/src/service/finalized_state.rs [hunk 3]
```diff
@@ -39,16 +43,36 @@ pub struct FinalizedState {
...
-            rocksdb::ColumnFamilyDescriptor::new("sapling_nullifiers", db_options.clone()),
-            rocksdb::ColumnFamilyDescriptor::new("orchard_nullifiers", db_options.clone()),
+            rocksdb::ColumnFamilyDescriptor::new("hash_by_height", reject_updates.clone()),
+            rocksdb::ColumnFamilyDescriptor::new("height_by_hash", reject_updates.clone()),
+            rocksdb::ColumnFamilyDescriptor::new("block_by_height", reject_updates.clone()),
+            rocksdb::ColumnFamilyDescriptor::new("tx_by_hash", reject_updates.clone()),
+            rocksdb::ColumnFamilyDescriptor::new("utxo_by_outpoint", reject_updates.clone()),
+            rocksdb::ColumnFamilyDescriptor::new("sprout_nullifiers", reject_updates.clone()),
+            rocksdb::ColumnFamilyDescriptor::new("sapling_nullifiers", reject_updates.clone()),
+            rocksdb::ColumnFamilyDescriptor::new("orchard_nullifiers", reject_updates),
```
Comment (teor2345) [inline zebra-state/src/service/finalized_state.rs]: TODO: document how the code implements the consensus rules:

> A nullifier MUST NOT repeat either within a transaction, or across transactions in a valid blockchain. Sprout and Sapling and Orchard nullifiers are considered disjoint, even if they have the same bit pattern.

https://zips.z.cash/protocol/protocol.pdf#nullifierset

> A transaction is not valid if it would have added a nullifier to the nullifier set that already exists in the set

https://zips.z.cash/protocol/protocol.pdf#commitmentsandnullifiers

**Comment by teor2345** [comment]
The macOS failure is unrelated, it looks like #2163.

Most of the first ~30 peers we tried failed, and took up the entire timeout. That will get better when each initial handshake is in its own task.

File: zebra-state/src/service/finalized_state.rs [hunk 4]
```diff
@@ -39,16 +43,36 @@ pub struct FinalizedState {
...
+            // value_pools allows Update merges:
+            // chain value pool balances get updated for each block
+            //rocksdb::ColumnFamilyDescriptor::new("value_pools", db_options.clone()),
```
Comment (jvff) [inline zebra-state/src/service/finalized_state.rs]: Question: this will be uncommented in a separate PR, right? Would it make sense to have a `TODO` here, or is it going to be uncommented soon?

File: zebra-state/src/service/finalized_state/merge_operator.rs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    _key: &[u8],
+    existing_val: Option<&[u8]>,
+    operands: &mut MergeOperands,
+) -> Option<Vec<u8>> {
+    // reject overwrites to keys that already have values
+    if existing_val.is_some() {
+        return None;
+    }
+
+    if operands.size_hint().0 > 1 {
```
Comment (jvff) [inline zebra-state/src/service/finalized_state/merge_operator.rs]: Warning: Using `size_hint` might be a little dangerous. This will currently work because the bounds are exact in [the source code](https://docs.rs/rocksdb/0.16.0/src/rocksdb/merge_operator.rs.html#216), and I think there's no reason to expect this to change in the future.

However, if it does change in the future, the `== 0` and the `== 1` cases might be incorrect. For example, the bounds could be `0..` (`(0, None)`) or `1..3` (`(1, Some(3))`).  It might be more cautious to do something like:

```
let value_to_insert = operands.next()?;

if operands.next().is_some() {
    // reject partial merges and combined writes of multiple keys
    None
} else {
    // if there is a single value, write it
    Some(value_to_insert)
}
```

**Review by jvff — CHANGES_REQUESTED** [review]
Looks good, but I did add a warning note about possibly making the code more cautious of any future `rocksdb` changes. As far as I can tell, this is just an extra precaution, and it would still work if left as is :+1:

File: book/src/dev/rfcs/0005-state-updates.md [hunk 6]
```diff
@@ -600,20 +600,39 @@ order on byte strings is the numeric ordering).
...
-| `height_by_hash`     | `block::Hash`         | `BE32(height)`                      |
-| `block_by_height`    | `BE32(height)`        | `Block`                             |
-| `tx_by_hash`         | `transaction::Hash`   | `(BE32(height) \|\| BE32(tx_index))`|
-| `utxo_by_outpoint`   | `OutPoint`            | `TransparentOutput`                 |
-| `sprout_nullifiers`  | `sprout::Nullifier`   | `()`                                |
-| `sapling_nullifiers` | `sapling::Nullifier`  | `()`                                |
-| `orchard_nullifiers` | `orchard::Nullifier`  | `()`                                |
-| `sprout_anchors`     | `sprout::tree::Root`  | `()`                                |
-| `sapling_anchors`    | `sapling::tree::Root` | `()`                                |
+| Tree                  | Keys                  | Values                               | Merge  |
```
Comment (teor2345) [inline book/src/dev/rfcs/0005-state-updates.md]:
```suggestion
| Column Family         | Keys                  | Values                               | Merge  |
```

File: book/src/dev/rfcs/0005-state-updates.md (hunk 2 above)
Comment (dconnolly) [inline book/src/dev/rfcs/0005-state-updates.md]:
```suggestion
| `orchard_incremental` | `BE32(height)`        | `orchard::tree::NoteCommitmentTree`            | Delete |

```

File: book/src/dev/rfcs/0005-state-updates.md [hunk 7]
```diff
@@ -600,20 +600,39 @@ order on byte strings is the numeric ordering).
...
+| `block_by_height`     | `BE32(height)`        | `Block`                              | Never  |
+| `tx_by_hash`          | `transaction::Hash`   | `(BE32(height) \|\| BE32(tx_index))` | Never  |
+| `utxo_by_outpoint`    | `OutPoint`            | `transparent::Output`                | Delete |
+| `sprout_nullifiers`   | `sprout::Nullifier`   | `()`                                 | Never  |
+| `sapling_nullifiers`  | `sapling::Nullifier`  | `()`                                 | Never  |
+| `orchard_nullifiers`  | `orchard::Nullifier`  | `()`                                 | Never  |
+| `sprout_anchors`      | `sprout::tree::Root`  | `()`                                 | Never  |
+| `sprout_incremental`  | *TODO*                | *TODO*                               | *TODO* |
+| `sapling_anchors`     | `sapling::tree::Root` | `()`                                 | Never  |
+| `sapling_incremental` | *TODO*                | *TODO*                               | *TODO* |
```
Comment (dconnolly) [inline book/src/dev/rfcs/0005-state-updates.md]:
```suggestion
| `sapling_incremental` | `BE32(height)`        | `sapling::tree::NoteCommitmentTree`            | Delete |
```

File: book/src/dev/rfcs/0005-state-updates.md [hunk 8]
```diff
@@ -600,20 +600,39 @@ order on byte strings is the numeric ordering).
...
+| `orchard_anchors`     | `orchard::tree::Root` | `()`                                 | Never  |
+| `orchard_incremental` | *TODO*                | *TODO*                               | *TODO* |
+| `history_incremental` | `BE32(height)`        | `zcash_history::NodeData`            | Delete |
+| `value_pools`         | `()`                  | `ValueBalance<NonNegative>`          | Update |
```
Comment (teor2345) [inline book/src/dev/rfcs/0005-state-updates.md]:
```suggestion
| `history_incremental` | `BE32(height)` (1)    | `zcash_history::NodeData`            | Delete |
| `value_pools`         | `()`                  | `ValueBalance<NonNegative>`          | Update |

(1): The history incremental tree for the finalized tip consists of `NodeData` for a small number of peaks.
     We map peak indexes to heights. Peaks are written once, then deleted when they are no longer required.
```

File: book/src/dev/rfcs/0005-state-updates.md [hunk 9]
```diff
@@ -600,20 +600,39 @@ order on byte strings is the numeric ordering).
...
+| `hash_by_height`      | `BE32(height)`        | `block::Hash`                        | Never  |
+| `height_by_hash`      | `block::Hash`         | `BE32(height)`                       | Never  |
+| `block_by_height`     | `BE32(height)`        | `Block`                              | Never  |
+| `tx_by_hash`          | `transaction::Hash`   | `(BE32(height) \|\| BE32(tx_index))` | Never  |
+| `utxo_by_outpoint`    | `OutPoint`            | `transparent::Output`                | Delete |
+| `sprout_nullifiers`   | `sprout::Nullifier`   | `()`                                 | Never  |
+| `sapling_nullifiers`  | `sapling::Nullifier`  | `()`                                 | Never  |
+| `orchard_nullifiers`  | `orchard::Nullifier`  | `()`                                 | Never  |
+| `sprout_anchors`      | `sprout::tree::Root`  | `()`                                 | Never  |
+| `sprout_incremental`  | *TODO*                | *TODO*                               | *TODO* |
```
Comment (dconnolly) [inline book/src/dev/rfcs/0005-state-updates.md]: TODO: check that this is sufficient

```suggestion
| `sprout_incremental` | `BE32(height)`        | `sprout::tree::NoteCommitmentTree`           | Delete |
```

File: book/src/dev/rfcs/0005-state-updates.md (hunk 2 above)
Comment (teor2345) [reply book/src/dev/rfcs/0005-state-updates.md]: Overwrites can happen if there are no sprout/sapling/orchard transfers in a block.

File: zebra-state/src/service/finalized_state.rs (hunk 4 above)
Comment (teor2345) [reply zebra-state/src/service/finalized_state.rs]: I'll add a TODO when I do the revisions

File: zebra-state/src/service/finalized_state/merge_operator.rs (hunk 5 above)
Comment (teor2345) [reply zebra-state/src/service/finalized_state/merge_operator.rs]: Yes I noticed, it should really implement `ExactSizeIterator`.

In the meantime you're right, we should code defensively.

File: zebra-state/src/service/finalized_state.rs (hunk 4 above)
Comment (teor2345) [reply zebra-state/src/service/finalized_state.rs]: 86f6ab34

File: zebra-state/src/service/finalized_state/merge_operator.rs (hunk 5 above)
Comment (teor2345) [reply zebra-state/src/service/finalized_state/merge_operator.rs]: eccaaeb7

File: zebra-state/src/service/finalized_state.rs (hunk 3 above)
Comment (teor2345) [reply zebra-state/src/service/finalized_state.rs]: 8437f210

**Comment by teor2345** [comment]
This PR got a lot more complicated than I expected.

The code works, but the check is in the wrong place in the finalized state.

I need to move the check to the non-finalized state, and probably split the UTXO and nullifier changes.

File: zebra-state/src/service/finalized_state.rs (hunk 4 above)
Comment (teor2345) [reply zebra-state/src/service/finalized_state.rs]: This change is obsolete, because the initial design didn't work.

File: zebra-state/src/service/finalized_state/merge_operator.rs (hunk 5 above)
Comment (teor2345) [reply zebra-state/src/service/finalized_state/merge_operator.rs]: This change is obsolete, because the initial design didn't work.

**Comment by teor2345** [comment]
I'm going to open a PR that obsoletes this PR some time over the next day or two.
